src/core/provider.cc

```diff
@@ -698,15 +744,39 @@ InferResponseProvider::AllocateOutputBuffer(
 
   void* buffer = nullptr;
   void* buffer_userp = nullptr;
-
+#ifdef TRTIS_ENABLE_GPU
+  int current_device;
+  // Not checking error here as "no device" will be returned on CPU-only system.
+  // We still need to get device since provider doesn't know, however, the
+  // actual memory type returned from 'alloc_fn_' implies whether the provider
+  // should (can) recover current device.
+  cudaGetDevice(&current_device);
+#endif  // TRTIS_ENABLE_GPU
   TRTSERVER_Error* err = alloc_fn_(
       allocator_, &buffer, &buffer_userp, name.c_str(), alloc_byte_size,
-      preferred_memory_type, 0 /* region_id */, alloc_userp_);
+      preferred_memory_type, preferred_memory_type_id, alloc_userp_);
+  Status status;
+#ifdef TRTIS_ENABLE_GPU
+  // [TODO] change to check actual memory type after
+  // https://github.com/NVIDIA/tensorrt-inference-server/pull/780
+  if (preferred_memory_type == TRTSERVER_MEMORY_GPU) {
```

Comment:
I don't think we should actually guard this with preferred_memory_type since we don't know what happened during the alloc function. But it is ok for now since I think this code will be changing due to callback changes for mem_type_id.

Reply:
After some thinking, even after the allocation function improvement, we still can't rely on the actual memory type returned by the allocation function, because we should treat it as a black box and then the returned memory type doesn't necessarily tell whether the current device has been changed.

I think we should just always call cudaGet/SetDevice() regardless of their returned error code.